File: py/candlestick_patterns/patterns/three_outside.py
"""Three Outside Up/Down pattern (3-candle reversal)."""
from __future__ import annotations

from ..core.primitives import is_white, is_black, real_body
from ...fuzzy import t_product_all


def three_outside(self) -> float:
    """Three Outside Up/Down: a three-candle reversal pattern.

    Must have:
    - first and second candles form an engulfing pattern,
    - third candle confirms the direction by closing higher (up) or
      lower (down).

    Three Outside Up: first candle is black, second is white engulfing
    the first, third closes higher than the second.

    Three Outside Down: first candle is white, second is black engulfing
    the first, third closes lower than the second.

    Category C: both branches evaluated, return stronger signal.

    Returns:
        Continuous float in [-100, +100].
    """
    if not self._enough(3):
        return 0.0

    o1, h1, l1, c1 = self._bar(3)
    o2, h2, l2, c2 = self._bar(2)
    o3, h3, l3, c3 = self._bar(1)

    # Fuzzy engulfment width.
    eq_avg = self._avg(self._equal, 1) if hasattr(self, '_equal') else 0.0
    eq_width = self._fuzz_ratio * eq_avg if eq_avg > 0.0 else 0.0

    # Three Outside Up: black + white engulfing + 3rd closes higher.
    bull_signal = 0.0
    if is_black(o1, c1) and is_white(o2, c2):
        mu_enc_upper = self._mu_ge_raw(max(o2, c2), max(o1, c1), eq_width)
        mu_enc_lower = self._mu_lt_raw(min(o2, c2), min(o1, c1), eq_width)
        rb2 = real_body(o2, c2)
        width = self._fuzz_ratio * rb2 if rb2 > 0.0 else 0.0
        mu_close_higher = self._mu_gt_raw(c3, c2, width)
        conf = t_product_all(mu_enc_upper, mu_enc_lower, mu_close_higher)
        bull_signal = conf * 100.0

    # Three Outside Down: white + black engulfing + 3rd closes lower.
    bear_signal = 0.0
    if is_white(o1, c1) and is_black(o2, c2):
        mu_enc_upper = self._mu_ge_raw(max(o2, c2), max(o1, c1), eq_width)
        mu_enc_lower = self._mu_lt_raw(min(o2, c2), min(o1, c1), eq_width)
        rb2 = real_body(o2, c2)
        width = self._fuzz_ratio * rb2 if rb2 > 0.0 else 0.0
        mu_close_lower = self._mu_lt_raw(c3, c2, width)
        conf = t_product_all(mu_enc_upper, mu_enc_lower, mu_close_lower)
        bear_signal = -conf * 100.0

    if abs(bull_signal) >= abs(bear_signal):
        return bull_signal
    return bear_signal

File: ts/candlestick-patterns/patterns/three-outside.ts
/** Three Outside Up/Down pattern (3-candle reversal). */
import { CandlestickPatternsEngine } from '../core/engine.ts';
import { tProductAll } from '../../fuzzy/index.ts';
import { realBody, isWhite, isBlack } from '../core/primitives.ts';

/**
 * Three Outside Up/Down: a three-candle reversal pattern.
 *
 * Must have:
 * - first and second candles form an engulfing pattern,
 * - third candle confirms the direction by closing higher (up) or
 *   lower (down).
 *
 * Three Outside Up: first candle is black, second is white engulfing
 * the first, third closes higher than the second.
 *
 * Three Outside Down: first candle is white, second is black engulfing
 * the first, third closes lower than the second.
 *
 * Category C: both branches evaluated, return stronger signal.
 *
 * Returns:
 *     Continuous float in [-100, +100].
 */
export function threeOutside(cp: CandlestickPatternsEngine): number {
    if (!cp.enough(3)) return 0.0;

    const b1 = cp.bar(3);
    const b2 = cp.bar(2);
    const b3 = cp.bar(1);

    // Fuzzy engulfment width.
    const eqAvg = cp.avgCS(cp.equal, 1);
    let eqWidth = 0.0;
    if (eqAvg > 0.0) eqWidth = cp.fuzzRatio * eqAvg;

    // Three Outside Up: black + white engulfing + 3rd closes higher.
    let bullSignal = 0.0;
    if (isBlack(b1.o, b1.c) && isWhite(b2.o, b2.c)) {
        const muEncUpper = cp.muGeRaw(Math.max(b2.o, b2.c), Math.max(b1.o, b1.c), eqWidth);
        const muEncLower = cp.muLtRaw(Math.min(b2.o, b2.c), Math.min(b1.o, b1.c), eqWidth);
        const rb2 = realBody(b2.o, b2.c);
        let width = 0.0;
        if (rb2 > 0.0) width = cp.fuzzRatio * rb2;
        const muCloseHigher = cp.muGtRaw(b3.c, b2.c, width);
        bullSignal = tProductAll(muEncUpper, muEncLower, muCloseHigher) * 100.0;
    }

    // Three Outside Down: white + black engulfing + 3rd closes lower.
    let bearSignal = 0.0;
    if (isWhite(b1.o, b1.c) && isBlack(b2.o, b2.c)) {
        const muEncUpper = cp.muGeRaw(Math.max(b2.o, b2.c), Math.max(b1.o, b1.c), eqWidth);
        const muEncLower = cp.muLtRaw(Math.min(b2.o, b2.c), Math.min(b1.o, b1.c), eqWidth);
        const rb2 = realBody(b2.o, b2.c);
        let width = 0.0;
        if (rb2 > 0.0) width = cp.fuzzRatio * rb2;
        const muCloseLower = cp.muLtRaw(b3.c, b2.c, width);
        bearSignal = -tProductAll(muEncUpper, muEncLower, muCloseLower) * 100.0;
    }

    return Math.abs(bullSignal) >= Math.abs(bearSignal) ? bullSignal : bearSignal;
}
